text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1389 @@
+
+The protobuf definition of this service is:
+
+```protobuf
+service ArrowStreamService {
+  // The ArrowStream endpoint is a bi-directional stream used to send batch of `BatchArrowRecords` from the exporter
+  // to the collector. The collector returns `BatchStatus` messages to acknowledge the `BatchArrowRecords` messages received.
+  rpc ArrowStream(stream BatchArrowRecords) returns (stream BatchStatus) {}
+}
+```
+
+We use a stream-oriented protocol **to get rid of the overhead of specifying the schema and dictionaries for each
+batch.**
+A state will be maintained receiver side to keep track of the schemas and dictionaries.
+The [Arrow IPC format](#arrow-ipc-format) has been
+designed to follow this pattern and also allows the dictionaries to be sent incrementally.
+
+A `BatchArrowRecords` message is composed of 3 attributes. The protobuf definition is:
+
+```protobuf
+message BatchArrowRecords {
+  // [mandatory] Batch ID. Must be unique in the context of the stream.
+  string batch_id = 1;
+
+  // [mandatory] A collection of payloads containing the data of the batch.
+  repeated OtlpArrowPayload otlp_arrow_payloads = 2;
+
+  // [optional] Delivery type (BEST_EFFORT by default).
+  DeliveryType delivery_type = 3;
+}
+```
+
+The `batch_id` attribute is a unique identifier for the batch inside the scope of the current stream. It is used to
+uniquely
+identify the batch in the egress `BatchStatus` stream. See the [Batch Id generation](#batch-id-generation) section for
+more information on the implementation of this identifier.
+
+The `otlp_arrow_payloads` attribute is a list of `OtlpArrowPayload` messages. Each `OtlpArrowPayload` message represents
+a table of data encoded in a columnar format (metrics, logs, or traces). Although not used in the current version of
+this protocol, this attribute is an array to allow representing different types of entities.
+More details on the `OtlpArrowPayload` columns in the
+section [Mapping OTEL entities to Arrow records](#mapping-otel-entities-to-arrow-records).
+
+The `delivery_type` attribute is an optional attribute that can be used to indicate the type of message delivery
+guarantee expected by the sender. The delivery types are defined by the following protobuf enumeration:
+
+```protobuf
+enum DeliveryType {
+  BEST_EFFORT = 0;
+}
+```
+
+The delivery type `BEST_EFFORT` is the default value. This means that the sender expects the receiver to acknowledge
+receipt of a message and to do its best to pass the message on to the rest of the downstream chain.
+
+More specifically, an `OtlpArrowPayload` protobuf message is defined as:
+
+```protobuf
+message OtlpArrowPayload {
+  // [mandatory] A unique id assigned to a sub-stream of the batch sharing the same schema, and dictionaries.
+  string sub_stream_id = 1;
+
+  // [mandatory] Type of the OTLP Arrow payload.
+  OtlpArrowPayloadType type = 2;
+
+  // [optional] Serialized Arrow dictionaries
+  repeated EncodedData dictionaries = 3;
```

Comment:
I may have misused the Arrow IPC API. I'll come back to you on this point after revisiting this part of the code.

Reply:
> All in all this looks really well written and researched (plus already incorporated feedback I gave about the types), I just had a few comments here and there along with a couple questions about specific design choices. There's a few cases where it looks like you're essentially duplicating effort or adding more work for yourselves that the IPC format handles for you.
> 
> CC @lquerel

I'd like to give some context to this comment. First of all, a **huge thanks** to Matt (@zeroshade) for his numerous feedbacks that will help to improve and consolidate this project (spec and implementation). 

**Matt is one of the main maintainer of the Apache Arrow Go project**. I invited him to do a review of this OTEP from the perspective of an Arrow expert, which he accepted to do without any problem. Some of his comments have led to changes in the OTEP, others will require some improvements or explorations in the reference implementation.  I don't expect these changes to take long to implement. Thanks again for the review.

Reply:
It's "funny" because after reviewing the code, I just had to remove the field in the proto specification because it was not used in the project. So we can conclude that the use of Arrow IPC is ok :)

See this diff https://github.com/f5/otel-arrow-adapter/pull/90/files